言及機能追加

ItohYuto opened the pull request:
- 日報の作成、編集、削除時に言及について更新される。
- 日報の保存処理及び言及の保存処理は同じトランザクション内で実行しどちらかで失敗した場合、処理をロールバックする。
- 日報が自身を言及することはできない。
- 言及されている日報詳細ページでは言及している日報のリンクの一覧が表示される。

言及機能を追加したことによりReportControllerのCreateアクション、Updateアクションの実装が複雑かつ長くなったので簡潔にできないかと試行錯誤しました。
複雑かつ同じような処理がある箇所はメソッド化したりなどしました。
メソッド化したもの
- 日報の本文内からメンションする日報を見つける処理。
(app/controllers/reports_controller.rb: find_mention_reports)
- メンションを追加する処理
(app/models/report.rb: line:23, menntions)
- メンションを削除する処理
(app/models/report.rb: line:27, menntion_cancels)

ReportControllerのCreateアクション、Updateアクションに関して、何かご意見いただけるとありがたいです。

ItohYuto committed:
言及機能追加

日報の作成、編集、削除時に言及について更新される。
日報の保存処理及び言及の保存処理は同じトランザクション内で実行しどちらかで失敗した場合、処理をロールバックする。
日報が自身を言及することはできない。
言及されている日報詳細ページでは言及している日報のリンクの一覧が表示される。

JunichiIto commented on app/controllers/reports_controller.rb:
```diff
@@ -20,16 +20,33 @@ def edit; end
 
   def create
     @report = current_user.reports.new(report_params)
+    mention_reports = find_mention_reports(report_params[:content])
```
fat controllerにならないように設計を見直してみてください〜。
https://qiita.com/jnchito/items/2b57316b39d92ca0e8dc

ItohYuto replied:
scaffoldに合わせる形でコード実装しました。
そのため、Reportモデル内に処理を移動しております。

JunichiIto commented on app/models/mention_relationship.rb:
```diff
@@ -0,0 +1,9 @@
+  belongs_to :report
+  belongs_to :mention, class_name: 'Report'
+
+  validates :report_id, presence: true
```
belongs_to は自動的に必須チェックが走るのでバリデーションは不要です

ItohYuto replied:
修正しました。

JunichiIto commented on app/models/mention_relationship.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class MentionRelationship < ApplicationRecord
```
relationship = 関連、という意味なので、関連するデータに「なんとか関連」という名前のモデルがいると用語が錯綜してややこしくなります。
以下の記事などを参考にしてもっとよい命名を検討してみてください。
https://qiita.com/tkawa/items/dc3e313021f32fd91ca6

ItohYuto replied:
ReportMentionに修正しました。

JunichiIto commented on app/models/mention_relationship.rb:
```diff
@@ -0,0 +1,9 @@
+
+class MentionRelationship < ApplicationRecord
+  belongs_to :report
+  belongs_to :mention, class_name: 'Report'
```
片方は元のモデルと同じ名前（report）で、もう片方が違う名前というのはちょっとわかりづらいですね。
それぞれにわかりやすい別名を付けた方が良いと思います。

ItohYuto replied:
mention_to(～をメンションする)、mentioned_by(～にメンションされている)に修正しました。

JunichiIto commented on app/models/report.rb:
```diff
@@ -7,11 +7,27 @@ class Report < ApplicationRecord
 
+  has_many :mention_relationships, dependent: :destroy
+  has_many :mentioning_reports, through: :mention_relationships, source: :mention
+  has_many :reverse_of_relationships, class_name: 'MentionRelationship', foreign_key: 'mention_id', inverse_of: 'mention', dependent: :destroy
```
app/models/mention_relationship.rb のコメントと同様です。

> 片方は元のモデルと同じ名前（report）で、もう片方が違う名前というのはちょっとわかりづらいですね。
それぞれにわかりやすい別名を付けた方が良いと思います。

あと、"reverse_of_relationships" というのは直訳すると「逆の関連」になりますが、もし将来他にも同じようなモデルを追加する必要が出てきたら、それも "reverse_of_relationships" という名前を付けたくなってしまいそうです。
つまりちょっと抽象度が高すぎるのでもうちょっと具体性のある名前を付けた方がよいですね。

https://bootcamp.fjord.jp/articles/78

ItohYuto replied:
forward_mentions(前向きの言及)、backward_mentions(後ろ向きの言及)に修正しました。

JunichiIto commented on app/models/report.rb:
```diff
@@ -7,11 +7,27 @@ class Report < ApplicationRecord
     created_at.to_date
   end
+
+  def mentions(target_reports)
```
メソッドは原則として動詞で始めてください。
https://qiita.com/jnchito/items/459d58ba652bf4763820

ItohYuto replied:
メンション関連のメソッドに関しては以下のように命名を修正しました。
add_menntions
delete_menntions
create_menntions
update_mentions

JunichiIto commented on app/models/report.rb:
```diff
@@ -7,11 +7,27 @@ class Report < ApplicationRecord
   end
+
+  def mentions(target_reports)
+    target_reports.map { |report| mention_relationships.find_or_create_by(mention_id: report.id) if self != report }.compact
```
mapの中でデータが作られる（createされる）というのは、「驚きの大きい副作用」なので避けましょう。
https://blog.jnito.com/entry/2023/07/12/115733

ItohYuto replied:
修正しました。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -2,6 +2,20 @@
+    <div>
+      <strong>この記事は以下の記事で言及されています。</strong>
+      <ul>
+        <% @report.mentioned_reports.each do |report| %>
```
orderを指定した方がいいですね
https://qiita.com/jnchito/items/90c28c3f3e856add0a82

ItohYuto replied:
修正しました。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -2,6 +2,20 @@
+      <ul>
+        <% @report.mentioned_reports.each do |report| %>
+        <li>
+          <a>
```
link_to自体がaタグを生成するのでこのaタグは不要です。
（そもそもhref属性のないaタグはあまり意味がないです）

ItohYuto replied:
修正しました。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -2,6 +2,20 @@
+        <% @report.mentioned_reports.each do |report| %>
+        <li>
+          <a>
+            <%= link_to  report.title, report_path(report) %>
```
```suggestion
            <%= link_to report.title, report_path(report) %>
```

スペースが多いです

ItohYuto replied:
修正しました。

JunichiIto commented on db/migrate/20231218065809_create_mention_relationships.rb:
```diff
@@ -0,0 +1,9 @@
+class CreateMentionRelationships < ActiveRecord::Migration[7.0]
+  def change
+    create_table :mention_relationships do |t|
+      t.references :report, foreign_key: true
```
null禁止の制約も付けた方がいいですね

ItohYuto replied:
修正しました。

JunichiIto commented on test/fixtures/mention_relationships.yml:
```diff
@@ -0,0 +1,11 @@
+# column: value
+#
+two: {}
+# column: value
```
使わないファイルは削除しましょう

ItohYuto replied:
修正しました。

JunichiIto reviewed (commented):
コメントしました〜。

ItohYuto committed:
言及関連のテーブル名、モデル名、カラム名を変更
外部参照キーにnull禁止制約を追加
不要なバリデートを削除
mapメソッド内でデータを作らないように変更。
controllerをscaffoldに似た形になるように修正。
不要なaタグの削除
言及リンクの表示の際に表示順が一意になるように変更。
不要なスペース削除。
不要なファイルの削除

ItohYuto commented:
修正しましたので再度レビューお願いいたします。

## 日報作成時の動き
1. 日報新規作成
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/85178d0b-e1c5-432e-8dea-d22b4619bfbd)
2. 言及されている日報の詳細ページ（２ページ分並べて表示しています。）
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/24f06b0c-4303-41a2-a211-006feae59444)
3. 複数のに峰から言及されている場合の日報詳細ページ
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/310b371b-d4d8-4fa2-902a-d36cf9e09f9f)
4. ターミナルの出力
一つのトランザクション内(画像赤枠内)で`Report Create`,`ReportMention Create`が実行されている。
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/097c239e-2200-495a-a3c7-db7af31add05)
## 日報編集時の動き
1. 日報編集
日報新規作成時の動きの1.日報新規作成で作成した日報を編集しています。
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/44b6fc64-6b58-4698-a426-e67be6997ffb)
2. 言及削除された日報
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/032b535b-d8d5-462f-896c-98ca62a0f0af)
3. 変更の無い日報 
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/afa53376-2ec1-423d-93b0-00217f81fd89)
4. 編集時に追加された日報
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/c7419e37-0d93-4c38-8110-e4e05b203430)
5. ターミナル出力
一つのトランザクション内(画像赤枠内)で`Report Update`,`ReportMention Delete All`,`ReportMention Create`,が実行されている。
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/211fb375-59ff-41e0-9cf9-8bb134ba7753)

## 日報削除時の動き
削除用の日報を新規作成して、言及されている日報詳細ページにリンクが存在していることを確認
1. 削除前の状態
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/519620fb-cca8-463c-acd8-db0fef2c9fed)
2. 日報削除後
言及されていた日報からリンクが消えていることを確認
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/df356c01-222a-4526-9d85-95c1c1fe7934)
3. ターミナルの出力
一つのトランザクション内(画像赤枠内)で`ReportMention Destroy`,`Report Destroy`が実行されている。
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/8d6129c4-8172-4168-9866-54bb8e9cb931)

## rubocop
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/d7ad5804-6d0a-4808-a702-b6e72258e470)

## erblint
![image](https://github.com/ItohYuto/fjord-books_app-2023/assets/88817643/8427a713-398b-4da5-ba8e-fb75749e4ef6)

JunichiIto commented on app/models/report.rb:
```diff
@@ -7,11 +7,45 @@ class Report < ApplicationRecord
+  private
+
+  def find_mention_reports
+    report_ids = content.scan(%r{http://localhost:3000/reports/[0-9]*}).map { |url| url.delete_prefix('http://localhost:3000/reports/') }
```
`([0-9]+)`のように正規表現のキャプチャ機能を使うとシンプルになるかもしれません。
あと、`*`は0件以上なので、`reports/` にもマッチします。`+`にして1件以上にした方がベターです。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -2,6 +2,18 @@
 
 <div class="show-item">
   <%= render @report %>
+  <% if !@report.mentioned_reports.empty? %>
```
Railsだと、

```suggestion
  <% if @report.mentioned_reports.present? %>
```

こう書くことが多いかも。

もしくは

```suggestion
  <% unless @report.mentioned_reports.empty? %>
```

とかでしょうか

JunichiIto reviewed (commented):
コメントしました〜。

ItohYuto committed:
日報本文からURLを探す際に、正規表現のキャプチャを使用するように変更。
否定演算子を使用しない形に変更。
